frontend/catalyst/jax_tracer.py

```diff
@@ -1191,3 +1213,410 @@ def is_leaf(obj):
+                for i, tape in enumerate(tapes):
+                    qdevice_p.bind(
+                        rtd_lib=device.backend_lib,
+                        rtd_name=device.backend_name,
+                        rtd_kwargs=str(device.backend_kwargs),
+                    )
+                    # breakpoint()
+                    qreg_in = qalloc_p.bind(len(device.wires))
+                    # If the program is batched, that means that it was transformed.
+                    # If it was transformed, that means that the program might have
+                    # changed the output. See `split_non_commuting`
+                    if qnode_transformed or device_modify_measurements:
+                        # TODO: In the future support arbitrary output from the user function.
+                        output = tape.measurements
+                        _, trees = jax.tree_util.tree_flatten(output, is_leaf=is_leaf)
+                    else:
+                        output = return_values_flat
+                        trees = return_values_tree
+
+                    mcm_config = qnode.execute_kwargs["mcm_config"]
+                    qrp_out = trace_quantum_operations(
+                        tape, device, qreg_in, ctx, trace, mcm_config
+                    )
+                    meas, meas_trees = trace_quantum_measurements(device, qrp_out, output, trees)
+                    qreg_out = qrp_out.actualize()
+
+                    meas_tracers = [trace.full_raise(m) for m in meas]
+                    meas_results = tree_unflatten(meas_trees, meas_tracers)
+
+                    # TODO: Allow the user to return whatever types they specify.
+                    if qnode_transformed or device_modify_measurements:
+                        assert isinstance(meas_results, list)
+                        if len(meas_results) == 1:
+                            transformed_results.append(meas_results[0])
+                        else:
+                            transformed_results.append(tuple(meas_results))
+                    else:
+                        transformed_results.append(meas_results)
+
+                    # Deallocate the register after each tape
+                    qdealloc_p.bind(qreg_out)
+
+                    # Get the jaxpr for this tape and clear the tracing frame for next tape
+                    # breakpoint()
+                    if type(transformed_results[i]) is not tuple:
+                        out_tracers = [transformed_results[i]]
+                    else:
+                        out_tracers = [_ for _ in transformed_results[i]]
+                    jp = EvaluationContext.find_jaxpr_frame().to_jaxpr(out_tracers)
+
+                    # EvaluationContext.find_jaxpr_frame().eqns.clear()
+                    EvaluationContext.find_jaxpr_frame().eqns.append(
+                        new_jaxpr_eqn([], [], jax.core.Primitive("_tape_cut"), [], no_effects)
+                    )
```

Comment:
What is this tape_cut primitive?

Reply:
It is a custom primitive to separate different tapes. The tracing frame needs to be continuous across all the tapes otherwise some constval/constvar would be lost.